### PR TITLE
Bal 684 collect directors images in collection flow

### PR DESCRIPTION
**Pull request by chesterkmr**
### Description
- Added director document inputs to definition
- Added director documents blocks in backoffice
- Various bugfixes
- Fixed missing disabled state on Multiselect and checkboxes
- UI Bump

### Examples and references
 *Document fields on director form item*

image

*Directors block in backoffice*

image

*Document selection in revision popup*

image

*Allowing to edit only revised fields by disabling other fields*
image

### CODE REVIEW

File: apps/backoffice-v2/.vscode/launch.json
Comment (Omri-Levy) [inline apps/backoffice-v2/.vscode/launch.json:1]: Why was this file deleted?

File: apps/backoffice-v2/src/pages/Entity/components/CallToAction/CallToAction.tsx [hunk 1]
```diff
@@ -45,23 +59,37 @@ export const CallToAction: FunctionComponent<ICallToActionProps> = ({ value, dat
...
+  });
+
+  const documentPickerProps = useDocumentSelection(documentSelection, resetReasonAndComment);
+
+  const handleDialogClose = useCallback(
+    (open: boolean) => {
+      if (!open) onDialogClose && onDialogClose();
+    },
+    [onDialogClose],
+  );
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/pages/Entity/components/CallToAction/CallToAction.tsx]: Move into `useCallToActionLogic`.

File: apps/backoffice-v2/src/pages/Entity/components/CallToAction/components/DocumentPicker/DocumentPicker.tsx
Comment (Omri-Levy) [inline apps/backoffice-v2/src/pages/Entity/components/CallToAction/components/DocumentPicker/DocumentPicker.tsx:1]: Maybe this should be a generic component under molecules with the label and so on passed as props?

File: apps/backoffice-v2/src/pages/Entity/components/Case/hooks/useCaseActionsLogic/useCaseActionsLogic.tsx [hunk 2]
```diff
@@ -14,6 +14,8 @@ import { useFilterId } from '../../../../../../common/hooks/useFilterId/useFilte
 import { useRevisionCaseMutation } from '../../../../../../domains/workflows/hooks/mutations/useRevisionCaseMutation/useRevisionCaseMutation';
 import { useCaseDecision } from '../useCaseDecision/useCaseDecision';
 import { tagToBadgeData } from '../../consts';
+import { selectDirectorsDocuments } from '../../../../hooks/useTasks/selectors/selectDirectorsDocuments';
+import { AnyObject } from '@ballerine/ui';
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/pages/Entity/components/Case/hooks/useCaseActionsLogic/useCaseActionsLogic.tsx:18]: Make sure Storybook still works after importing from `@ballerine/ui`. Otherwise we have other sources for this type.

File: apps/backoffice-v2/src/pages/Entity/components/Details/Details.tsx [hunk 3]
```diff
@@ -25,9 +32,11 @@ export const Details: FunctionComponent<IDetailsProps> = ({ id, value, hideSepar
         workflowId={workflow?.id}
         id={id}
         valueId={value?.id}
-        documents={workflow?.context?.documents}
+        documents={documents ?? workflow?.context?.documents}
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/pages/Entity/components/Details/Details.tsx]: This change will cause confusion and might result in unexpected outcomes. Let's at least receive documents as props for both cases.

File: apps/backoffice-v2/src/pages/Entity/components/DirectorsCallToAction/DirectorsCallToAction.tsx [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  ...restProps
+}) => {
+  const { documents, workflow, onReset, ...restValueProps } = value?.props || {};
+
+  const [documentId, setDocumentId] = useState<string | null>(null);
+
+  const documentsOptions: ICallToActionDocumentOption[] = useMemo(
+    () =>
+      documents.map(({ category, type, id }) => ({
+        name: toTitleCase(`${category as string} - ${type as string}`),
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/pages/Entity/components/DirectorsCallToAction/DirectorsCallToAction.tsx]: See instances of `valueOrNA`, otherwise instead of asserting make sure we have a default or at least an empty string over undefined.

File: apps/backoffice-v2/src/pages/Entity/components/DirectorsCallToAction/helpers.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  if (category === 'proof_of_identity_ownership' && type === 'selfie') {
+    return [
+      'Blurry image',
+      'Bad quality photo',
+      'Wrong document',
+      'Copy of a copy',
+      'Person in the selfie does not match ID',
+      'There was no person in the photo',
+    ];
+  }
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/pages/Entity/components/DirectorsCallToAction/helpers.ts:25]: This should be handled by schemas or the workflows service and not the backoffice.

File: apps/backoffice-v2/src/pages/Entity/hooks/useTasks/hooks/useDirectorsBlocks/useDirectorsBlocks.tsx [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+  }, []);
+
+  const blocks = useMemo(() => {
+    return directors
+      .filter(director => Array.isArray(director.additionalInfo?.documents))
+      .map(director => {
+        const { documents } = director.additionalInfo;
+
+        const multiDocumentsBlocks = documents.map((document, docIndex) => {
+          const isDoneWithRevision = document?.decision?.status === 'revised';
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/pages/Entity/hooks/useTasks/hooks/useDirectorsBlocks/useDirectorsBlocks.tsx:65]: This does not seem to align with the other instance of `isDoneWithRevision`.

File: apps/backoffice-v2/src/pages/Entity/hooks/useTasks/hooks/useDirectorsBlocks/useDirectorsBlocks.tsx [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+                            disabled:
+                              (!isDoneWithRevision && Boolean(document?.decision?.status)) ||
+                              noAction,
+                          },
+                        },
+                      },
+                ],
+              },
+              {
+                id: 'kyc-block',
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/pages/Entity/hooks/useTasks/hooks/useDirectorsBlocks/useDirectorsBlocks.tsx]: Make sure there are no references to unrelated blocks.

File: apps/backoffice-v2/src/pages/Entity/hooks/useTasks/hooks/useDirectorsBlocks/useDirectorsBlocks.tsx [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+                    type: 'container',
+                    value: [
+                      {
+                        type: 'subheading',
+                        value: `${valueOrNA(toTitleCase(document.category ?? ''))} - ${valueOrNA(
+                          toTitleCase(document.type ?? ''),
+                        )}`,
+                      },
+                      {
+                        title: 'Details test',
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/pages/Entity/hooks/useTasks/hooks/useDirectorsBlocks/useDirectorsBlocks.tsx]: Remove instances of `test`.

File: apps/backoffice-v2/src/pages/Entity/hooks/useTasks/hooks/useDirectorsBlocks/useDirectorsBlocks.tsx [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+                                pattern,
+                                dropdownOptions,
+                                value,
+                                formatMinimum,
+                                formatMaximum,
+                              },
+                            ]) => {
+                              const fieldValue = value || (document.properties?.[title] ?? '');
+                              const isDoneWithRevision = document?.decision?.status === 'revised';
+                              const isEditableDecision =
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/pages/Entity/hooks/useTasks/hooks/useDirectorsBlocks/useDirectorsBlocks.tsx]: Is editing the decision allowed now?

File: apps/backoffice-v2/src/pages/Entity/hooks/useTasks/hooks/useDirectorsBlocks/useDirectorsBlocks.tsx [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+                  },
+                ],
+              },
+            ],
+          };
+        });
+
+        const isDocumentRevision = documents.some(
+          document => document?.decision?.status === 'revision',
+        );
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/pages/Entity/hooks/useTasks/hooks/useDirectorsBlocks/useDirectorsBlocks.tsx:213]: Note we have helpers for extracting decision status.

File: apps/backoffice-v2/src/pages/Entity/hooks/useTasks/useTasks.tsx [hunk 11]
```diff
@@ -260,20 +280,40 @@ export const useTasks = ({
...
+                text: isLegacyReject ? 'Reject' : 'Re-upload needed',
+                props: {
+                  revisionReasons:
+                    workflow?.workflowDefinition?.contextSchema?.schema?.properties?.documents?.items?.properties?.decision?.properties?.revisionReason?.anyOf?.find(
+                      ({ enum: enum_ }) => !!enum_,
+                    )?.enum,
+                  rejectionReasons:
+                    workflow?.workflowDefinition?.contextSchema?.schema?.properties?.documents?.items?.properties?.decision?.properties?.rejectionReason?.anyOf?.find(
+                      ({ enum: enum_ }) => !!enum_,
+                    )?.enum,
```
Comment (Omri-Levy) [inline apps/backoffice-v2/src/pages/Entity/hooks/useTasks/useTasks.tsx:333]: Move to a variable and re-use below.

File: apps/kyb-app/jest.config.ts [hunk 12]
```diff
@@ -1,5 +1,9 @@
 const config = {
   setupFilesAfterEnv: ['<rootDir>/tests/setup.js'],
+  moduleNameMapper: {
+    '^@app(.*)$': '<rootDir>/src/$1',
+  },
```
Comment (Omri-Levy) [inline apps/kyb-app/jest.config.ts:5]: `kyb-app` should not have Jest as it uses Vitest. Second, we no longer use `@app` and use `@/` instead.

File: apps/kyb-app/src/components/organisms/UIRenderer/elements/JSONForm/components/CheckboxList/CheckboxList.tsx [hunk 13]
```diff
@@ -8,14 +9,14 @@ interface CheckboxListOption {
...
-  const { uiSchema, formData = [], onChange } = props;
+  const { uiSchema, formData = [], onChange, disabled } = props;
 
   const options = useMemo(() => {
     return (uiSchema?.['options'] as CheckboxListOption[]) || [];
   }, [uiSchema]);
 
   return (
-    <div className="flex flex-col gap-4">
+    <div className={clsx('flex flex-col gap-4', { 'pointer-events-none opacity-50': disabled })}>
```
Comment (Omri-Levy) [inline apps/kyb-app/src/components/organisms/UIRenderer/elements/JSONForm/components/CheckboxList/CheckboxList.tsx]: Use `cn` over `clsx` since it includes `tailwind-merge`.

File: apps/kyb-app/src/components/organisms/UIRenderer/elements/JSONForm/components/DocumentField/DocumentField.tsx [hunk 14]
```diff
@@ -1,117 +1,131 @@
...
-        // @ts-ignore
-        const createdFile = new File([''], file.fileNameInBucket || file.fileNameOnDisk, {
-          type: 'text/plain',
-        });
-
-        collectionFlowFileStorage.registerFile(fileId, createdFile);
-        setFile(createdFile);
+    if (persistedFile) return;
+
+    void fetchFile(fileId).then(file => {
```
Comment (Omri-Levy) [inline apps/kyb-app/src/components/organisms/UIRenderer/elements/JSONForm/components/DocumentField/DocumentField.tsx:95]: Why does this fetch request breaks out of `react-query`? There seems to be no error handling or loading handling either.

File: apps/kyb-app/src/components/organisms/UIRenderer/elements/JSONForm/components/DocumentField/DocumentField.tsx [hunk 15]
```diff
@@ -121,20 +135,81 @@ export const DocumentField = (
...
+
+      const context = stateApi.getContext();
+      //@ts-ignore
+      const documents = (get(context, pathToDocumentsList) as Document[]) || [];
+
+      let document = documents.find(
+        document =>
+          document &&
+          //@ts-ignore
+          document.id === serializeDocumentId(definition.options.documentData.id, inputIndex),
```
Comment (Omri-Levy) [inline apps/kyb-app/src/components/organisms/UIRenderer/elements/JSONForm/components/DocumentField/DocumentField.tsx]: Is both sides being undefined an option?
```suggestion
          //@ts-ignore
          document?.id === serializeDocumentId(definition.options.documentData.id, inputIndex),
```

File: apps/kyb-app/src/components/organisms/UIRenderer/elements/JSONForm/components/FileUploaderField/hooks/useFileAssigner/useFileAssigner.ts [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+    if (!ref.current) return;
+
+    if (!ref.current.files?.length) {
+      const files = new DataTransfer();
+
+      if (!(file instanceof File)) return;
+
+      files.items.add(file);
+      ref.current.files = files.files;
+    }
```
Comment (Omri-Levy) [inline apps/kyb-app/src/components/organisms/UIRenderer/elements/JSONForm/components/FileUploaderField/hooks/useFileAssigner/useFileAssigner.ts]:
```suggestion
    if (!ref.current || ref.current.files?.length) return;

      const files = new DataTransfer();

      if (!(file instanceof File)) return;

      files.items.add(file);
      ref.current.files = files.files;
```

File: apps/kyb-app/src/components/organisms/UIRenderer/elements/StepperUI/StepperUI.tsx [hunk 17]
```diff
@@ -5,53 +5,54 @@ import { VerticalLayout } from '@/components/atoms/Stepper/layouts/Vertical';
...
+    context: CollectionFlowContext;
   }) => {
-    if (!!Object.keys(pageError).length && currentPage.number === page.number) return 'warning';
     if (
-      uiElement?.isCompleted ||
-      page.number <=
-        // @ts-ignore
-        initialPageNumber?.current
+      Object.keys(pageError).length &&
+      Object.values(pageError).some(error => error.type === 'warning')
```
Comment (Omri-Levy) [inline apps/kyb-app/src/components/organisms/UIRenderer/elements/StepperUI/StepperUI.tsx]: If there are no keys there won't be any errors of type warning right? 🙂 
```suggestion
      Object.values(pageError ?? {}).some(error => error.type === 'warning')
```

File: apps/kyb-app/tests/setup.js [hunk 18]
```diff
@@ -1,9 +1,11 @@
...
 import { cleanup } from '@testing-library/react';
-import { afterEach } from 'vitest';
+import { afterEach, vi } from 'vitest';
 
 // runs a cleanup after each test case (e.g. clearing jsdom)
 afterEach(() => {
   cleanup();
 });
+
+global.jest = vi;
```
Comment (Omri-Levy) [inline apps/kyb-app/tests/setup.js:11]: Could you please explain this?

File: services/workflows-service/prisma/migrations/20230723221706_add_jsonb_deep_merge_with_options/migration.sql [hunk 19]
```diff
@@ -98,4 +98,4 @@ RETURNS jsonb AS $$
   };
 
   return deepMergeWithOptions(obj1, obj2, array_merge_option);
-$$ LANGUAGE plv8 IMMUTABLE STRICT;
\ No newline at end of file
+$$ LANGUAGE plv8 IMMUTABLE STRICT;
```
Comment (Omri-Levy) [inline services/workflows-service/prisma/migrations/20230723221706_add_jsonb_deep_merge_with_options/migration.sql:101]: What caused this change?

File: services/workflows-service/prisma/migrations/20231122094527_fix_jsonb_deep_merge_with_options_primitives_bug/migration.sql
Comment (Omri-Levy) [inline services/workflows-service/prisma/migrations/20231122094527_fix_jsonb_deep_merge_with_options_primitives_bug/migration.sql:1]: Can you explain this change?

File: services/workflows-service/scripts/workflows/dynamic-ui-workflow.ts [hunk 20]
```diff
@@ -386,6 +387,7 @@ export const generateDynamicUiWorkflow = async (prismaClient: PrismaClient, proj
     data: kybDynamicExample,
   });
 
+  //@ts-ignore
   await generateDynamicUiTest(prismaClient, workflow.id, projectId || workflow.projectId);
```
Comment (Omri-Levy) [inline services/workflows-service/scripts/workflows/dynamic-ui-workflow.ts]: Note if you're up to date with dev the project ID from DB should never be null and only one of these project ids need to be passed..

File: services/workflows-service/src/workflow/workflow.service.ts [hunk 21]
```diff
@@ -832,10 +843,14 @@ export class WorkflowService {
 
     const updatedWorkflow = await this.updateContextById(
       workflowId,
-      {
-        documents: [newDocument],
-      },
+      this.updateDocumentInContext(runtimeData.context, newDocument, documentsUpdateContextMethod),
       [projectId],
+      documentsUpdateContextMethod === 'director' ? 'by_index' : 'by_id',
```
Comment (Omri-Levy) [inline services/workflows-service/src/workflow/workflow.service.ts:848]: Maybe we should change `contextUpdateMethod` from `base` | `director` to `by_index` | `by_id`. Unless there are more differences beyond the plv8 merge method `by_index` could be potentially be used by other places than directors. Make sure we get this past @alonp99 though.

File: services/workflows-service/src/workflow/workflow.service.ts [hunk 22]
```diff
@@ -887,17 +900,80 @@ export class WorkflowService {
...
+  private getDocuments(
+    context: WorkflowRuntimeData['context'],
+    documentsUpdateContextMethod: 'base' | 'director' = 'base',
+  ) {
+    switch (documentsUpdateContextMethod) {
+      case 'base':
+        return context.documents;
+      case 'director':
+        return this.getDirectorsDocuments(context);
+    }
```
Comment (Omri-Levy) [inline services/workflows-service/src/workflow/workflow.service.ts]: Add a default case please. 🙏🏼

File: apps/backoffice-v2/.vscode/launch.json
Comment (chesterkmr) [reply apps/backoffice-v2/.vscode/launch.json:1]: It was added by mistake and should not be there. For cases when somebody will decide to use VSCode workspace settings then this folder should be added to .gitignore.

File: apps/backoffice-v2/src/pages/Entity/components/CallToAction/components/DocumentPicker/DocumentPicker.tsx
Comment (chesterkmr) [reply apps/backoffice-v2/src/pages/Entity/components/CallToAction/components/DocumentPicker/DocumentPicker.tsx:1]: I don't think so. Component is used just once and in single place. Main idea behind this component is to wrap chunk of code to reduce size of `CallToAction` component where its been used.

File: apps/backoffice-v2/src/pages/Entity/components/Case/hooks/useCaseActionsLogic/useCaseActionsLogic.tsx (hunk 2 above)
Comment (chesterkmr) [reply apps/backoffice-v2/src/pages/Entity/components/Case/hooks/useCaseActionsLogic/useCaseActionsLogic.tsx:18]: Storybook works.

File: apps/backoffice-v2/src/pages/Entity/components/DirectorsCallToAction/helpers.ts (hunk 5 above)
Comment (chesterkmr) [reply apps/backoffice-v2/src/pages/Entity/components/DirectorsCallToAction/helpers.ts:25]: This is sounds like additional feature to me. In order to do so we have to update schemas and add reasons to schemas in common package. Consider open task for it.

File: apps/backoffice-v2/src/pages/Entity/hooks/useTasks/hooks/useDirectorsBlocks/useDirectorsBlocks.tsx (hunk 9 above)
Comment (chesterkmr) [reply apps/backoffice-v2/src/pages/Entity/hooks/useTasks/hooks/useDirectorsBlocks/useDirectorsBlocks.tsx]: Bad naming

File: apps/kyb-app/src/components/organisms/UIRenderer/elements/JSONForm/components/DocumentField/DocumentField.tsx (hunk 14 above)
Comment (chesterkmr) [reply apps/kyb-app/src/components/organisms/UIRenderer/elements/JSONForm/components/DocumentField/DocumentField.tsx:95]: Query is not needed in this case. Logic is different. Regarding error handling you are right, but in this case handle of this error supposed to display toast and thats it. Toasts not in app yet, so no handling.

File: services/workflows-service/prisma/migrations/20230723221706_add_jsonb_deep_merge_with_options/migration.sql (hunk 19 above)
Comment (chesterkmr) [reply services/workflows-service/prisma/migrations/20230723221706_add_jsonb_deep_merge_with_options/migration.sql:101]: Just checked dev, there no change, could be formatting ?

**Review by Omri-Levy — APPROVED** [review]
